components/homme/src/theta-l/eos.F90

```diff
@@ -275,64 +323,72 @@ subroutine get_dirk_jacobian(JacL,JacD,JacU,dt2,dp3d,phi_i,pnh,exact,&
+             (( -dphi(:,:,k) )*(1-kappa)*dp3d_i(:,:,k))
 
-           JacD(k,:,:) = 1+(dt2*g)**2 *(pnh(:,:,k)/((phi_i(:,:,k)-phi_i(:,:,k+1))*(1-kappa)) &
-             +pnh(:,:,k-1)/( (phi_i(:,:,k-1)-phi_i(:,:,k))*(1-kappa)))/dp3d_i(:,:,k)
+           JacD(k,:,:) = 1+(dt2*g)**2 *(pnh(:,:,k)/(( -dphi(:,:,k) )*(1-kappa)) &
+             +pnh(:,:,k-1)/( ( -dphi(:,:,k-1) )*(1-kappa)))/dp3d_i(:,:,k)
 
         end if
       end do
     else ! use finite difference approximation to Jacobian with differencing size espie
-      ! compute Jacobian of F(phi) = phi +const + (dt*g)^2 *(1-dp/dpi) column wise
+      ! compute Jacobian of F(dphi) = phi +const + (dt*g)^2 *(1-dp/dpi) column wise
+      ! if NEWTON_DPHI is defined, compute Jacobian of:
+      !                   dF(dphi) = dphi +const + (dt*g)^2 *d(1-dp/dpi) 
       ! we only form the tridagonal entries and this code can easily be modified to
       ! accomodate sparse non-tridigonal and dense Jacobians, however, testing only
       ! the tridiagonal of a Jacobian is probably sufficient for testing purpose
       do k=1,nlev
         e=0
         e(:,:,k)=1
-        phi_i_temp(:,:,:) = phi_i(:,:,:)
-        phi_i_temp(:,:,k) = phi_i(:,:,k) + epsie*e(:,:,k)
+        dphi_temp(:,:,:) = dphi(:,:,:)
+        ! dphi_tmp(k)= ( phi(k+1)-(phi(k)+eps ) = dphi(k)-eps
+        ! dphi_tmp(k-1)= ( phi(k)+eps-(phi(k-1) ) = dphi(k-1)+eps
+        dphi_temp(:,:,k) = dphi(:,:,k) - epsie*e(:,:,k)
+        if (k>1) dphi_temp(:,:,k-1) = dphi(:,:,k-1) + epsie*e(:,:,k)
+
         if (theta_hydrostatic_mode) then
-          dpnh_dp_i_epsie(:,:,:)=1.d0
+           !dpnh_dp_i_epsie(:,:,:)=1.d0
+           delta_mu=0
         else
-         call pnh_and_exner_from_eos(hvcoord,vtheta_dp,dp3d,phi_i_temp,pnh,exner,dpnh_dp_i_epsie)
+           call pnh_and_exner_from_eos2(hvcoord,vtheta_dp,dp3d,dphi_temp,pnh,exner,dpnh_dp_i_epsie,'get_dirk_jacobian')
+           delta_mu(:,:,:)=(g*dt2)**2*(dpnh_dp_i(:,:,:)-dpnh_dp_i_epsie(:,:,:))/epsie
         end if
+
+        JacD(k,:,:) = 1 +  delta_mu(:,:,k)
         if (k.eq.1) then
-          JacL(k,:,:) = (g*dt2)**2 * (dpnh_dp_i(:,:,k+1)-dpnh_dp_i_epsie(:,:,k+1))/epsie
-          JacD(k,:,:) = 1 + (g*dt2)**2 * (dpnh_dp_i(:,:,k)-dpnh_dp_i_epsie(:,:,k))/epsie
+           JacL(k,:,:) =   delta_mu(:,:,k+1)
         elseif (k.eq.nlev) then
-          JacD(k,:,:)   = 1 + (g*dt2)**2 * (dpnh_dp_i(:,:,k)-dpnh_dp_i_epsie(:,:,k))/epsie
-          JacU(k-1,:,:) = (g*dt2)**2 * (dpnh_dp_i(:,:,k-1)-dpnh_dp_i_epsie(:,:,k-1))/epsie
+           JacU(k-1,:,:) = delta_mu(:,:,k-1)
         else
-          JacL(k,:,:)   = (g*dt2)**2 * (dpnh_dp_i(:,:,k+1)-dpnh_dp_i_epsie(:,:,k+1))/epsie
-          JacD(k,:,:)   = 1 + (g*dt2)**2 * (dpnh_dp_i(:,:,k)-dpnh_dp_i_epsie(:,:,k))/epsie
-          JacU(k-1,:,:) = (g*dt2)**2 * (dpnh_dp_i(:,:,k-1)-dpnh_dp_i_epsie(:,:,k-1))/epsie
+           JacL(k,:,:)   = delta_mu(:,:,k+1)
+           JacU(k-1,:,:) = delta_mu(:,:,k-1)
         end if
+
```

Comment:
Above, for inexact jacobian:
k==1: JacL(1) is set
k==nlev: JacU(71) is set
2<k<nlev: JacL(2:71), JacU(1:70) are set
so, all indices are accounted for, but it is not readable. delta_mu is precomputed, so lines above could be simplified.